# Project: ML - Logistic Regression, Random Forests (Lending Club)

<br>

## Questions to address:
- Predict the outcome of loans given by the lending club to inform a conservative investor
- Dataset: lending club - stats on lenders and boorrowers (approved and declined loan applications)

<br>

## Tools:
- Models: 
  - Logistic Regression
  - Random Forests  
- Feature Preparation, Selection and Engineering (transforming and processing):
  - identify columns that leak information about the future, don't affect borrower's ability to pay the loan, need to be clean up, require a lot of processing, contain redundant information
  - drop columns with only one unique value (after removing nan)
  - handle missing values:
    - remove columns with more than 1% missing values, keep employment length as it is likely a good predictor
    - drop rows with missing values
  - identify numerical needing conversion and extraneous columns
  - dummy coding for categorical features
    - transform numerical to categorical (for those where numbers have no meaning)
    - identify text columns to make categorical:
      - only those with only a few unique values
      - remove low varianve columns (more than 95% of the values are the same)
    - deal with NaNs, then dummy code
  - identify and clean the target colum (lookout for class inbalance)  
- Error Metric: TPR and FPR  
- K-fold Validation and predictions

<br>

### load defaults

In [3]:
import numpy as np
import pandas as pd
import seaborn as sns
import re
import requests 

%matplotlib inline
import matplotlib.pyplot as plt
from matplotlib.ticker import MultipleLocator
from matplotlib import rcParams
import matplotlib.dates as mdates
from datetime import datetime
from IPython.display import display, Math

from functions import *

plt.rcParams.update({'axes.titlepad': 20, 'font.size': 12, 'axes.titlesize':20})

colors = [(0/255,107/255,164/255), (255/255, 128/255, 14/255), 'red', 'green', '#9E80BA', '#8EDB8E', '#58517A']
Ncolors = 10
color_map = plt.cm.Blues_r(np.linspace(0.2, 0.5, Ncolors))
#color_map = plt.cm.tab20c_r(np.linspace(0.2, 0.5, Ncolors))


#specific to this project
from sklearn.feature_selection import RFECV

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import KFold, cross_val_predict
from sklearn.ensemble import RandomForestClassifier

print("Defaults Loaded")

Defaults Loaded


<br>

## Dataset: Lending Club stats

In [8]:
loans = pd.read_csv('./data/loans_2007.csv', low_memory=False)
display(loans.iloc[:3,:13])
print("Number of Columns: {:d}".format(len(loans.columns)))

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501  1296599.0     5000.0       5000.0           4975.0   36 months   
1  1077430  1314167.0     2500.0       2500.0           2500.0   60 months   
2  1077175  1313524.0     2400.0       2400.0           2400.0   36 months   

  int_rate  installment grade sub_grade emp_title emp_length home_ownership  
0   10.65%       162.87     B        B2       NaN  10+ years           RENT  
1   15.27%        59.83     C        C4     Ryder   < 1 year           RENT  
2   15.96%        84.33     C        C5       NaN  10+ years           RENT

Number of Columns: 52


<br>

### Feature Preparation, Selection and Engineering:
- identify and clean the target colum (lookout for class inbalance)
- identify columns that:
  - leak information about the future (after he loan was funded)
  - don't affect the borrower's ability to pay back the loan
  - formatted poorly and need to be clean up
  - require more data or a lot of processing
  - contain redundant information
- drop columns with only one unique value (after removing nan)
- handle missing values:
  - remove columns with more than 1% missing values, keep employment length as it is likely a good predictor
  - drop rows with missing values
- identify numerical needing conversion and extraneous columns
- dummy coding for categorical features
  - transform numerical to categorical (for those where numbers have no meaning)
  - identify text columns to make categorical:
    - only those with only a few unique values
    - remove low varianve columns (more than 95% of the values are the same)
  - deal with NaNs, then dummy code

**target column**
- loan_status is the only column that describes if the loan was paid in time, delayed or defaulted
- it contains text values and needs to be converted to numerical

In [9]:
print(loans['loan_status'].value_counts())

Fully Paid                                             33136
Charged Off                                             5634
Does not meet the credit policy. Status:Fully Paid      1988
Current                                                  961
Does not meet the credit policy. Status:Charged Off      761
Late (31-120 days)                                        24
In Grace Period                                           20
Late (16-30 days)                                          8
Default                                                    3
Name: loan_status, dtype: int64


only fully paid and charged off characterize loans with a final outcome (the others are ongoing)
- select only rows wih loan_status - fully paid or charged off
- assign fully paid to 1 and charged of to 0, then we can have a binary classification problem

In [10]:
sel = (loans['loan_status']== 'Fully Paid') | (loans['loan_status']== 'Charged Off')
loans = loans[sel]

mapping_dict = {
    "loan_status": {
        "Charged Off": 0,
        "Fully Paid": 1
    }
}
loans = loans.replace(mapping_dict)
print(loans['loan_status'].value_counts())

1    33136
0     5634
Name: loan_status, dtype: int64


**class inbalance in the target column**, keep in mind

clean features

In [11]:
cols_to_drop = ['id', 'member_id', 'funded_amnt', 'funded_amnt_inv', 'grade', 'sub_grade', 'emp_title', 
                'issue_d', 'zip_code', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 
                'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'recoveries', 
                'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt']
loans.drop(cols_to_drop, axis=1, inplace=True)
print(len(loans.columns))


#drop columns with only one unique value (after removing nan)
drop_columns = []

for element in loans.columns:
    if(len(loans[element].dropna().unique())==1):
        drop_columns.append(element)
        
loans.drop(drop_columns, axis=1, inplace=True)
print(len(loans.columns))

#remove columns with more than 1% missing values (keep employment length as it is likely a good predictor)
null_counts = loans.isnull().sum()
#print(null_counts[null_counts>len(loans)*0.01])
print(null_counts[null_counts>0.])
loans.drop('pub_rec_bankruptcies', axis=1, inplace=True)

#drop rows with missing values
loans.dropna(axis=0, inplace=True)
null_counts = loans.isnull().sum()
#print(null_counts[null_counts>0])

#write file
loans.to_csv('./data/filtered_loans_2007.csv', index=False)

32
23
emp_length              1036
title                     11
revol_util                50
last_credit_pull_d         2
pub_rec_bankruptcies     697
dtype: int64


identify numerical needing conversion and extraneous columns

In [12]:
loans = pd.read_csv('./data/filtered_loans_2007.csv')
object_columns_df = loans.select_dtypes(include=['object'])
display(object_columns_df[:1])

term int_rate emp_length home_ownership verification_status  \
0   36 months   10.65%  10+ years           RENT            Verified   

       purpose     title addr_state earliest_cr_line revol_util  \
0  credit_card  Computer         AZ         Jan-1985      83.7%   

  last_credit_pull_d  
0           Jun-2016

In [13]:
#convert to numeric:
loans['int_rate'] = loans['int_rate'].str.rstrip('%').astype(float)
loans['revol_util'] = loans['revol_util'].str.rstrip('%').astype(float)

#remove extraneous
cols_to_drop = ['last_credit_pull_d', 'earliest_cr_line']
loans.drop(cols_to_drop, axis=1, inplace=True)

**dummy coding**:
- home_ownership, verification_status, emp_length, term: few discrete values; encode as dummy
- purpose, title: overlaping information; keep purpose since title contains 18881 discrete values
- addr_state: 50 different values; remove

In [14]:
#cols = ['home_ownership', 'verification_status', 'emp_length', 'term', 'addr_state', 'purpose','title']

#for element in cols:    
#    print(len(object_columns_df[element].value_counts()))
#    print(object_columns_df[element].value_counts()[:5])
#    print('\n')

#drop 'addr_state' and 'title'
cols_to_drop = ['addr_state', 'title']
loans.drop(cols_to_drop, axis=1, inplace=True)

#map emp_length
mapping_dict = {
    "emp_length": {
        "10+ years": 10,
        "9 years": 9,
        "8 years": 8,
        "7 years": 7,
        "6 years": 6,
        "5 years": 5,
        "4 years": 4,
        "3 years": 3,
        "2 years": 2,
        "1 year": 1,
        "< 1 year": 0,
        "n/a": 0
    }
}
loans = loans.replace(mapping_dict)

#dummy code 'home_ownership', 'verification_status', 'purpose' and 'term'
cols = ['home_ownership', 'verification_status', 'purpose', 'term']
for element in cols:
    loans[element] = loans[element].astype('category')

dummy_df = pd.get_dummies(loans[cols])
loans = pd.concat([loans, dummy_df], axis=1)
loans.drop(cols, axis=1, inplace=True)

print(len(loans.columns))


#write file
loans.to_csv('./data/cleaned_loans_2007.csv', index=False)

38


<br>

### Error Metric
- mostly concerned with false positives (cost money) and false negatives (loose potential money)
- because of class inbalance a classifier can predict one for every row and still have high accuracy, use false positives and negatives
- optimize for high recall (true positive rate) and low fall-out (false positive rate):
  - $TPR=\frac{\mathrm{True Positives}}{\mathrm{True Positives+False Negatives}}$
  - $FPR=\frac{\mathrm{False Positives}}{\mathrm{False Positives+True Negatives}}$

<br>

### Logistic Regression Model with kfold validation and predictions
- good first model for binary classification problems:
  - quick to train and we can iterate more quickly
  - less prone to overfitting than more complex models like decision trees
  - easy to interpret

In [27]:
loans = pd.read_csv('./data/cleaned_loans_2007.csv')

features = loans.drop('loan_status', axis=1)
target = loans['loan_status']

lr = LogisticRegression(solver='lbfgs', max_iter=3000)
kf = KFold(10, random_state=1)

predictions = cross_val_predict(lr, features, target, cv=kf)
predictions = pd.Series(predictions)

tp = len(predictions[(target==1) & (predictions==1)])
tn = len(predictions[(target==0) & (predictions==0)])
fp = len(predictions[(target==0) & (predictions==1)])
fn = len(predictions[(target==1) & (predictions==0)])

tpr = tp/(tp+fn)
fpr = fp/(fp+tn)

print("TPR:{:0.3f}, FPR:{:0.3f}".format(tpr, fpr))

TPR:0.999, FPR:0.996


**correct for inbalance** by penalizing misclassifications of the less prevalent class more than the other class:
- class_weight='balanced'

In [28]:
lr = LogisticRegression(solver='lbfgs', class_weight='balanced', max_iter=3000)
#cross validation accross all the rows of the training data (kfold=n, leave one out validation)
#kf = KFold(features.shape[0], random_state=1)
kf = KFold(10, random_state=1)
predictions = cross_val_predict(lr, features, target, cv=kf, )
predictions = pd.Series(predictions)

tp = len(predictions[(target==1) & (predictions==1)])
tn = len(predictions[(target==0) & (predictions==0)])
fp = len(predictions[(target==0) & (predictions==1)])
fn = len(predictions[(target==1) & (predictions==0)])

tpr = tp/(tp+fn)
fpr = fp/(fp+tn)

print("TPR:{:0.3f}, FPR:{:0.3f}".format(tpr, fpr))

TPR:0.533, FPR:0.331


setting class_wight to balanced assign a penalty of ~5.89 for misclassifying 0 (there as 5.89 times more 1's than 0's)
- we can **set the penalty manually** to try to lower the false positive rates (higher penalty for misclassifying the negative class)

In [29]:
penalty = {0: 10, 1: 1}
lr = LogisticRegression(solver = 'lbfgs', class_weight=penalty, max_iter=3000)
#kf = KFold(features.shape[0], random_state=1)
kf = KFold(10, random_state=1)
predictions = cross_val_predict(lr, features, target, cv=kf, )
predictions = pd.Series(predictions)

tp = len(predictions[(target==1) & (predictions==1)])
tn = len(predictions[(target==0) & (predictions==0)])
fp = len(predictions[(target==0) & (predictions==1)])
fn = len(predictions[(target==1) & (predictions==0)])

tpr = tp/(tp+fn)
fpr = fp/(fp+tn)

print("TPR:{:0.3f}, FPR:{:0.3f}".format(tpr, fpr))

TPR:0.191, FPR:0.073


lower false positive rate at the expense of true positives

<br>

### Random Forests Model with RFECV and kfold validation and predictions

**Feature Selection with RFECV**

In [39]:
def select_features(df, target, model):    
    #select numeric and drop NaNs
    df_new = df.select_dtypes([np.number]).dropna(axis=1)
    #drop survived and ID
    all_X = df_new.drop(target,axis=1)
    all_y = df_new[target]
        
    #cv is the number of folds
    selector = RFECV(model, cv=10)
    selector.fit(all_X, all_y)  
    optimized_columns = list(all_X.columns[selector.support_])
    
    print("Best Columns \n"+"-"*12+"\n{}\n".format(optimized_columns))
    
    return optimized_columns

penalty = {0: 50, 1: 1}
model = RandomForestClassifier(n_estimators=50, random_state=1, min_samples_leaf=5, class_weight=penalty)
target = 'loan_status'
optimized_columns = select_features(loans, target, model)

Best Columns 
------------
['loan_amnt', 'int_rate', 'installment', 'emp_length', 'annual_inc', 'dti', 'open_acc', 'revol_bal', 'revol_util', 'total_acc']



**RandomForestClassifier with RFECV features**

In [41]:
penalty = {0: 500, 1: 1}
model = RandomForestClassifier(n_estimators=150, random_state=1, min_samples_leaf=5, class_weight=penalty)
#kf = KFold(features.shape[0], random_state=1)
kf = KFold(10, random_state=1)
features = loans[optimized_columns]
target = loans['loan_status']
predictions = cross_val_predict(model, features, target, cv=kf, )
predictions = pd.Series(predictions)

tp = len(predictions[(target==1) & (predictions==1)])
tn = len(predictions[(target==0) & (predictions==0)])
fp = len(predictions[(target==0) & (predictions==1)])
fn = len(predictions[(target==1) & (predictions==0)])

tpr = tp/(tp+fn)
fpr = fp/(fp+tn)

print("TPR:{:0.3f}, FPR:{:0.3f}".format(tpr, fpr))

TPR:0.850, FPR:0.675


**RandomForestClassifier with all features**

In [49]:
penalty = {0: 500, 1: 1}
model = RandomForestClassifier(n_estimators=150, random_state=1, min_samples_leaf=5, class_weight=penalty)
#kf = KFold(features.shape[0], random_state=1)
kf = KFold(10, random_state=1)
features = loans.drop('loan_status', axis=1)
target = loans['loan_status']
predictions = cross_val_predict(model, features, target, cv=kf, )
predictions = pd.Series(predictions)

tp = len(predictions[(target==1) & (predictions==1)])
tn = len(predictions[(target==0) & (predictions==0)])
fp = len(predictions[(target==0) & (predictions==1)])
fn = len(predictions[(target==1) & (predictions==0)])

tpr = tp/(tp+fn)
fpr = fp/(fp+tn)

print("TPR:{:0.3f}, FPR:{:0.3f}".format(tpr, fpr))

TPR:0.764, FPR:0.517


Using only RFECV features has a higher TPR (0.850, in comparison with 0.764 for all features), but at the expense of a higher FPR (0.675, in comparison with 0.517 for all features). Which features to use will depend on the type of investor (more or less conservative)